### PR TITLE
datasource: improve metadata handling

### PR DESCRIPTION
**Pull request by crawford**

### CODE REVIEW

**Comment by crawford** [comment]
Testing in progress...

File: initialize/meta_data.go [hunk 1]
```diff
@@ -17,68 +17,46 @@
...
 
 import (
-	"encoding/json"
 	"sort"
 
 	"github.com/coreos/coreos-cloudinit/config"
+	"github.com/coreos/coreos-cloudinit/datasource"
 )
 
 // ParseMetaData parses a JSON blob in the OpenStack metadata service format,
```
Comment (crawford) [inline initialize/meta_data.go]: Not true.

File: initialize/meta_data.go [hunk 2]
```diff
@@ -17,68 +17,46 @@
...
-		Hostname            string            `json:"hostname"`
-		NetworkConfig       struct {
-			ContentPath string `json:"content_path"`
-		} `json:"network_config"`
-	}
-	if err := json.Unmarshal([]byte(contents), &metadata); err != nil {
-		return nil, err
-	}
-
+func ParseMetaData(metadata datasource.Metadata) *config.CloudConfig {
```
Comment (crawford) [inline initialize/meta_data.go]: This should be moved to another package.

File: initialize/meta_data.go [hunk 3]
```diff
@@ -17,68 +17,46 @@
...
 		}
 	}
 	cfg.Hostname = metadata.Hostname
-	cfg.NetworkConfigPath = metadata.NetworkConfig.ContentPath
-	return &cfg, nil
+	cfg.Internal.NetworkConfig = metadata.NetworkConfig
+	return &cfg
 }
 
 // ExtractIPsFromMetaData parses a JSON blob in the OpenStack metadata service
```
Comment (crawford) [inline initialize/meta_data.go]: Not true.

File: initialize/meta_data.go [hunk 4]
```diff
@@ -17,68 +17,46 @@
...
 // format and returns a substitution map possibly containing private_ipv4,
 // public_ipv4, private_ipv6, and public_ipv6 addresses.
-func ExtractIPsFromMetadata(contents []byte) (map[string]string, error) {
-	var ips struct {
-		PublicIPv4  string `json:"public-ipv4"`
-		PrivateIPv4 string `json:"local-ipv4"`
-		PublicIPv6  string `json:"public-ipv6"`
-		PrivateIPv6 string `json:"local-ipv6"`
+func ExtractIPsFromMetadata(metadata datasource.Metadata) map[string]string {
```
Comment (crawford) [inline initialize/meta_data.go]: This can probably be removed.

**Comment by crawford** [comment]
I can't figure out how GitHub is ordering these commits... Its not a standard ordering. Here's the real thing:

```
9605b5e datasource: remove FetchNetworkConfig step
42153ed initialize: change env to process from metadata
650a239 metadata: simplify merging of metadata
3e47c09 datasource: replace metadata map with struct
d4c617f config: standardize interface a bit
9441586 test: check all root golang files
be62a1d test: DRY out MockFilesystem
```

File: config/config.go [hunk 5]
```diff
@@ -27,14 +27,13 @@ import (
...
-	ManageEtcHosts    EtcHosts `yaml:"manage_etc_hosts"`
-	NetworkConfigPath string   `yaml:"-"`
-	NetworkConfig     string   `yaml:"-"`
+	SSHAuthorizedKeys []string  `yaml:"ssh_authorized_keys"`
+	CoreOS            CoreOS    `yaml:"coreos"`
+	WriteFiles        []File    `yaml:"write_files"`
+	Hostname          string    `yaml:"hostname"`
+	Users             []User    `yaml:"users"`
+	ManageEtcHosts    EtcHosts  `yaml:"manage_etc_hosts"`
+	Internal          Internals `yaml:"-"`
```
Comment (jonboulle) [inline config/config.go]: Seems like this could just be broken out entirely and passed (or its equivalent) as another arg to `Apply()` now

**Comment by jonboulle** [comment]
LGTM, please do that in follow up

img

**Comment by crawford** [comment]
Oh god, I can't hide it. I already added the commit to this PR. No going back ;)

**Comment by jonboulle** [comment]
:+1:

**Pull request merged**